### PR TITLE
upgrade lsp to 1.5

### PR DESCRIPTION
**Pull request by kokobd**
This is a prerequisite of #3046

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/3072"></a>

### CODE REVIEW

**Comment by kokobd** [comment]
I'm receiving a lot of logs like this if `--verbose` is enabled in `ghcide-tests`. And finally got OOM even with 64GB memory.
image

So disabling it for now

**Comment by kokobd** [comment]
I see. In `lsp`, there is a debug log, printing response body:
https://github.com/haskell/lsp/blob/8b63438313828a31ba1065ea865e504d532d093e/lsp/src/Language/LSP/Server/Control.hs#L224-L234
And in HLS, our logger sends a copy of log to LSP client:
https://github.com/haskell/haskell-language-server/blob/347a7187f20242540926b9927c59a45b18798c67/exe/Main.hs#L72-L82

Then it goes into infinite recursion... So we should not enable debug logs in `lsp` package itself.

**Comment by michaelpj** [comment]
Hmm, I thought I had convinced myself that wasn't a problem :| I'm not 100% sure what to do about this. It's nice to be able to log the responses from LSP, but we do indeed have a loop problem if we then forward them on like this.

**Comment by kokobd** [comment]
> It's nice to be able to log the responses from LSP

And some other tests depend on the logs, I had to fix them one by one. This makes the PR harder than I initially thought.

**Comment by kokobd** [comment]
Finally, the CI is almost green.

File: plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs [hunk 1]
```diff
@@ -33,7 +33,7 @@ provider _ideState typ contents fp _ = liftIO $ do
     config <- findConfigOrDefault file
     let (range, selectedContents) = case typ of
           FormatText    -> (fullRange contents, contents)
-          FormatRange r -> (r, extractRange r contents)
+          FormatRange r -> (normalize r, extractRange r contents)
```
Comment (fendor) [inline plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs:36]: Why this change? Is that a bug, where the fix doesn't actually belong here?

File: plugins/hls-rename-plugin/src/Ide/Plugin/Rename.hs [hunk 2]
```diff
@@ -48,10 +48,6 @@ import           Ide.Types
 import           Language.LSP.Server
 import           Language.LSP.Types
 
-instance Hashable Location
-instance Hashable Range
-instance Hashable Position
-instance Hashable UInt
```
Comment (fendor) [inline plugins/hls-rename-plugin/src/Ide/Plugin/Rename.hs:54]: Instances have been added to lsp-types-1.5. This is not a question or complaint, just a note for the next reviewer

File: hls-test-utils/src/Test/Hls/Util.hs [hunk 3]
```diff
@@ -322,38 +314,31 @@ expectNoMoreDiagnostics timeout doc src = do
...
+    when (timeout > 0) $
         -- Give any further diagnostic messages time to arrive.
         liftIO $ sleep timeout
         -- Send a dummy message to provoke a response from the server.
         -- This guarantees that we have at least one message to
         -- process, so message won't block or timeout.
-        void $ Test.sendNotification (SCustomMethod "non-existent-method") A.Null
-    handleMessages
+    testId <- Test.sendRequest (SCustomMethod "test") (A.toJSON GetFilesOfInterest)
+    handleMessages testId
```
Comment (fendor) [inline hls-test-utils/src/Test/Hls/Util.hs:324]: this sounds like it is going to perform some actual work (although very minor), seems dubious to me?

**Review by fendor — APPROVED** [review]
LGTM, outstanding work!

File: plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs (hunk 1 above)
Comment (kokobd) [inline plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs:36]: It won't work without this change. I don't know why. All other formatters use `normalize` already.

File: hls-test-utils/src/Test/Hls/Util.hs [hunk 4]
```diff
@@ -322,38 +314,31 @@ expectNoMoreDiagnostics timeout doc src = do
 -- If timeout is 0 it will wait until the session timeout
 waitForDiagnosticsFromSourceWithTimeout :: Seconds -> TextDocumentIdentifier -> String -> Test.Session [Diagnostic]
 waitForDiagnosticsFromSourceWithTimeout timeout document source = do
```
Comment (kokobd) [inline hls-test-utils/src/Test/Hls/Util.hs:316]: Previously, this function relied on logs to work. But the current test logger doesn't send logs to lsp client. I found a very similar function in `ghcide` which doesn't rely on logs, so changed this function accordingly.

File: plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs (hunk 1 above)
Comment (fendor) [reply plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs:36]: I assumed, floskell performed the normalisation internally. That, or no one is running the tests.

File: hls-test-utils/src/Test/Hls/Util.hs (hunk 4 above)
Comment (fendor) [reply hls-test-utils/src/Test/Hls/Util.hs:316]: Right, but why are you sending `(A.toJSON GetFilesOfInterest)`. Seems like `A.Null` would express our intent more clearly?

File: plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs (hunk 1 above)
Comment (kokobd) [reply plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs:36]: The test case "formats a range" failed.
I didn't find out the reason, but all other formatting plugins use `normalize`, and this fixed the problem.

File: hls-test-utils/src/Test/Hls/Util.hs (hunk 3 above)
Comment (kokobd) [reply hls-test-utils/src/Test/Hls/Util.hs:324]: I adjusted the indentation to 4 spaces so the diff is messier. This is previously `SCustomMethod "non-existent-method"`, and the client waits for a logging diagnostics from the server, which is no longer present. I think instead of adding the log back, change the test to not rely on a specific form of logger is better.

And there is a similar function in `ghcide` which is newer and doesn't rely on logs to function properly:
https://github.com/haskell/haskell-language-server/blob/347a7187f20242540926b9927c59a45b18798c67/ghcide/test/src/Development/IDE/Test.hs#L92-L101

File: hls-test-utils/src/Test/Hls/Util.hs (hunk 4 above)
Comment (kokobd) [reply hls-test-utils/src/Test/Hls/Util.hs:316]: Yes, `A.Null` is better, and is working!

File: plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs (hunk 1 above)
Comment (fendor) [reply plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs:36]: good enough for me

File: ghcide/test/exe/Main.hs [hunk 5]
```diff
@@ -6986,18 +6990,20 @@ positionMappingTests =
...
-                        pure (range, replacement, newPos)
-                forAll
-                    (suchThatMap gen
-                        (\(range, replacement, newPos) -> positionResultToMaybe $ (range, replacement, newPos,) <$> fromCurrent range replacement newPos)) $
-                    \(range, replacement, newPos, oldPos) ->
-                    toCurrent range replacement oldPos === PositionExact newPos
+                        pure (rope, range, replacement)
+                monadicIO $ forAllM gen
+                    $ \(rope, range, replacement) -> do
+                        newRope <- liftIO $ applyChange (toCologActionWithPrio $ cmapWithPrio LogVfs recorder) rope (TextDocumentContentChangeEvent (Just range) Nothing replacement)
```
Comment (michaelpj) [inline ghcide/test/exe/Main.hs:7001]: Why not pass `mempty :: LogAction Identity` here? then you don't need IO, and you'll just not get hte log messages which you don't care about anyway. Then I think you can keep the old code.

**Review by michaelpj — APPROVED** [review]

File: ghcide/src/Development/IDE/Types/Logger.hs [hunk 6]
```diff
@@ -381,3 +380,14 @@ priorityToLsp =
     Info    -> MtInfo
     Warning -> MtWarning
     Error   -> MtError
+
+toCologActionWithPrio :: (MonadIO m, HasCallStack) => Recorder (WithPriority msg) -> LogAction m (WithSeverity msg)
```
[the remark this answers is not in the document]
Comment (kokobd) [reply ghcide/src/Development/IDE/Types/Logger.hs:384]: This is the first time `co-log-core` was introduced in HLS, so I think there isn't a existing function for this.

File: ghcide/test/exe/Main.hs (hunk 5 above)
Comment (kokobd) [reply ghcide/test/exe/Main.hs:7001]: Good advice!
